pkg/sources/github/feedlet.go

```diff
@@ -0,0 +1,396 @@
+
+	id, err := strconv.ParseInt(webhookID, 10, 64)
+	if err != nil {
+		glog.Errorf("failed to convert webhook %q to int64 : %s", webhookID, err)
+		return err
+	}
+	_, err = client.Repositories.DeleteHook(ctx, owner, repo, id)
+	if err != nil {
+		if errResp, ok := err.(*ghclient.ErrorResponse); ok {
+			// If the webhook doesn't exist, nothing to do
+			if errResp.Message == "Not Found" {
+				glog.Errorf("webhook doesn't exist, nothing to delete.")
+				return nil
+			}
+		}
+		glog.Errorf("gailed to delete the webhook: %#v", err)
+		return err
+	}
+	glog.Infof("deleted webhook %q successfully", webhookID)
+	return nil
+}
+
+func (t *GithubEventSource) StartFeed(trigger sources.EventTrigger, target string) (*sources.FeedContext, error) {
+
+	// Create the Receive Adapter Service that will accept incoming requests from GitHub.
+	service, err := t.createReceiveAdapter(trigger, target)
+	if err != nil {
+		glog.Warningf("failed to create service: %v", err)
+		return nil, err
+	}
+
+	glog.Infof("created Service: %+v", service)
+
+	// Start watching the Receive Adapter Service for it's updated domain name. This will be passed
+	// to GitHub as part of the webhook registration.
+	receiveAdaptorDomain, err := t.waitForServiceDomain(service.GetObjectMeta().GetName())
+	if err != nil {
+		glog.Infof("failed to get the service: %v", err)
+	}
+
+	return t.createWebhook(trigger, receiveAdaptorDomain)
+}
+
+func (t *GithubEventSource) waitForServiceDomain(serviceName string) (string, error) {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	wt := watchTimeout
+	opts := metav1.ListOptions{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: "v1alpha1",
+		},
+		FieldSelector:  fmt.Sprintf("metadata.name=%s", serviceName),
+		LabelSelector:  "receive-adapter=github",
+		Watch:          true,
+		TimeoutSeconds: &wt,
+	}
+
+	w, err := sc.Watch(opts)
+	if err != nil {
+		glog.Infof("failed to create watch: %v", err)
+		return "", err
+	}
+
+	eventChan := w.ResultChan()
+	for {
+		event, more := <-eventChan
+		if !more {
+			glog.Infof("expected a Service object, but got %T", event.Object)
+			return "", fmt.Errorf("no more")
+		}
+		switch event.Type {
+		case watch.Error:
+			return "", fmt.Errorf("watched Service error")
+		case watch.Deleted:
+			return "", fmt.Errorf("watched Service deleted")
+		case watch.Added:
+			fallthrough
+		case watch.Modified:
+			service, ok := event.Object.(*v1alpha1.Service)
+			if !ok {
+				glog.Infof("expected a Service object, but got %T", event.Object)
+				continue
+			}
+			if service.Name == serviceName {
+				status := service.Status
+				if status.Domain != "" {
+					w.Stop()
+					glog.Infof("got domain: %s", event.Object)
+					return status.Domain, nil
+				}
+			}
+		}
+	}
+}
+
+func receiveAdapterName(resource string) string {
+	serviceName := fmt.Sprintf("%s-%s-%s", "github", resource, postfixReceiveAdapter) // TODO: this needs more UUID on the end of it.
```

Comment:
If you want a quick & dirty UID, we've already vendored a uuid library. uuid.New().String()